packages/react-dom/src/client/ReactDOMInput.js

```diff
@@ -212,35 +240,66 @@ export function postMountWrapper(
 ) {
   const node = ((element: any): InputWithWrapperState);
 
+  // Do not assign value if it is already set. This prevents user text input
+  // from being lost during SSR hydration.
   if (props.hasOwnProperty('value') || props.hasOwnProperty('defaultValue')) {
+    const type = props.type;
+    const isButton = type === 'submit' || type === 'reset';
+
     // Avoid setting value attribute on submit/reset inputs as it overrides the
     // default value provided by the browser. See: #12872
-    const type = props.type;
-    if (
-      (type === 'submit' || type === 'reset') &&
-      (props.value === undefined || props.value === null)
-    ) {
+    if (isButton && (props.value === undefined || props.value === null)) {
       return;
     }
 
-    const initialValue = toString(node._wrapperState.initialValue);
-    const currentValue = node.value;
```

Comment:
I inlined this value because it is only needed in one place in both feature flag branches.